### PR TITLE
[#311]  Feature/airdrop

### PR DESCRIPTION
**Pull request by ohaddahan**
Implementing https://github.com/metaplex-foundation/sugar/issues/311

### CODE REVIEW

**Comment by ohaddahan** [comment]
Ok so I tested it locally and it works @MarkSackerberg @maestroh1git .

Usage:
`target/debug/sugar mint --airdrop-list airdrop_list.json`

Will write:
`airdrop_results.json`

`airdrop_list.json` format is : 
```json
{
"address": "number"
}
```

`airdrop_results.json` format is : 
```json
"address": [
{
"signature": "sighash",
"status": "boolean"
}
]
```

@samuelvanderwaal code is a bit messy at the moment, will need to clean it a bit.
Testing wise, I just created a `devnet` CM and tested it.
Can do unit tests but I guess a similar test is needed, maybe with a test validator, this kind of setup already included in the repo?

**Comment by ohaddahan** [comment]
Can't say I'm super happy about the test I added, let me know if need to improve.

File: src/mint/process.rs [hunk 1]
```diff
@@ -95,10 +120,24 @@ pub async fn process_mint(args: MintArgs) -> Result<()> {
     };
     println!("\nMinting to {}", &receiver_pubkey);
 
-    let number = args.number.unwrap_or(1);
+    let number = match using_airdrop_list {
+        true => 0,
+        false => args.number.unwrap_or(1),
+    };
+
+    if number > 1 && using_airdrop_list {
```
Comment (samuelvanderwaal) [inline src/mint/process.rs:128]: This check will never get triggered.

**Review by samuelvanderwaal — COMMENTED** [review]
The logic seems to work but there are some formatting and output issues. For example, if I run `sugar mint --airdrop-list airdrop_list.json` I get the following result:

Screen Shot 2022-10-08 at 11 32 01 AM

It minted all the items correctly but says it minted to the wrong address and the progress bar displays 0/0. 

In general I feel that adding all this logic to minting makes it a bit messy but if the logic actually works and the code just needs some formatting and output clean up then it's probably ok. Otherwise, we could consider moving this to a separate command: `sugar airdrop` to make the logic simpler.

File: src/mint/airdrop/utils.rs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+pub fn load_airdrop_results(airdrop_list: &mut AirDropTargets) -> Result<AirDropResults> {
+    // Will load previous airdrop results from file and will also sync the results with the targets
+    let airdrop_results_path = Path::new("airdrop_results.json");
+    if !airdrop_results_path.exists() {
+        return Ok(AirDropResults::new());
+    }
+
+    let file = match File::open(airdrop_results_path) {
```
[the remark this answers is not in the document]
Comment (ohaddahan) [reply src/mint/airdrop/utils.rs:27]: Cool, thanks for the tip.

File: src/mint/process.rs (hunk 1 above)
Comment (ohaddahan) [reply src/mint/process.rs:128]: Yeah I was shoe horning the `number` and `airdrop` together, felt like it's getting messy.

**Comment by ohaddahan** [comment]
> The logic seems to work but there are some formatting and output issues. For example, if I run `sugar mint --airdrop-list airdrop_list.json` I get the following result:
> 
> Screen Shot 2022-10-08 at 11 32 01 AM
> 
> It minted all the items correctly but says it minted to the wrong address and the progress bar displays 0/0.
> 
> In general I feel that adding all this logic to minting makes it a bit messy but if the logic actually works and the code just needs some formatting and output clean up then it's probably ok. Otherwise, we could consider moving this to a separate command: `sugar airdrop` to make the logic simpler.

Yeah I tend to agree that a separate command will be better and will allow making quite a few things cleaner in the code.

File: src/mint/airdrop/utils.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ohaddahan) [reply src/mint/airdrop/utils.rs]: Applied.

**Comment by ohaddahan** [comment]
Progress bar fixed and separated into its own command.

**Comment by samuelvanderwaal** [comment]
I tried a simple airdrop with three addresses and a total of five NFTs. It appears to have worked successfully 
Screen Shot 2022-10-25 at 9 16 49 PM


but I got an error message saying 3/5 items failed to mint:

Screen Shot 2022-10-25 at 9 16 41 PM

I have tried this several times and the airdrop always seems to succeed but I always get an error message. 

Screen Shot 2022-10-25 at 9 24 45 PM

**Comment by samuelvanderwaal** [comment]
Airdrop results are only written when it succeeds. There should be a file written when it fails so the user can retry only the failed items, otherwise when a user tries to airdrop thousands of NFTs and a few of them fail they will have to do a lot of manual work to figure out which ones failed.

**Comment by samuelvanderwaal** [comment]
It seems to mostly work with just some weird edge cases. I'll try to do more testing tomorrow and will try it on a much larger batch.

**Comment by ohaddahan** [comment]
Thanks for the feedback, I'll give it another go and see if I can replicate your findings.

**Comment by ohaddahan** [comment]
> Airdrop results are only written when it succeeds. There should be a file written when it fails so the user can retry only the failed items, otherwise when a user tries to airdrop thousands of NFTs and a few of them fail they will have to do a lot of manual work to figure out which ones failed.
> 
> Edit: Ok, I see that `load_airdrop_results` syncs the airdrop results with the airdrop list, that makes sense.

I missed the early exit on failure, moved writing the results prior.

https://github.com/ohaddahan/sugar/blob/d39966f30d19198408af859725b4f9a5217cd55e/src/airdrop/process.rs#L163

**Comment by ohaddahan** [comment]
> I tried a simple airdrop with three addresses and a total of five NFTs. It appears to have worked successfully Screen Shot 2022-10-25 at 9 16 49 PM
> 
> but I got an error message saying 3/5 items failed to mint:
> 
> Screen Shot 2022-10-25 at 9 16 41 PM
> 
> I have tried this several times and the airdrop always seems to succeed but I always get an error message.
> 
> Screen Shot 2022-10-25 at 9 24 45 PM

Strange I can't reproduce a failure.
Also have https://github.com/ohaddahan/sugar/blob/feature/airdrop/tests/sugar-airdrop-test.sh that seems to run fine.

**Comment by MarkSackerberg** [comment]
Hey,
thanks for the PR!
I've done some testing today and I saw some points that we should change or at least discuss:
1. there is no default for --airdrop-list? Maybe we should add it as airdrop_list.json
2. We should add the airdrop-list format to the help command
3. If sugar can't find the airdrop list it just does 0 mints and prints "command successful". We should catch that and throw an error message instead
4. If sugar skips an airdrop because of the content of the results file we should print that. E.g. "Skipping mints for existing transactions in airdrop_results.json" or similar. At least it confused me that an airdrop of 4 NFTs showed 3/3 successul mints (this is because one was done and written to the results file beforehand)
5. Wallets are not checked for validity. Malformed wallet addresses result in a crash. E.g.: `thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: WrongSize', /home/ubuntu/metaplex/sugar/src/airdrop/process.rs:112:61`. Maybe we should validate them upfront and print out incorrect wallet addresses so that the user can check them?

Besides those minor things: Nice PR! Thank you very much!

**Comment by ohaddahan** [comment]
> 

Thanks for the feedback.

I think I got it all in https://github.com/metaplex-foundation/sugar/pull/375/commits/94c5f830ecf8b90dd2ce2dc8dd8bd57be57c11e1 .

**Comment by MarkSackerberg** [comment]
Awesome, thank you!
I've only found two minor things:
1. When mints are skipped there is a missing new line an I would change `Skipping mints 3 due to` to `Skipping 3 mints due to` instead
```bash
sugar airdrop
Skipping mints 3 due to existing transactions in airdrop_results.json[1/2] 🔍 Loading candy machine
Candy machine ID: 7V37SuLRQjLGhsHMbgmdeSjhoLpSKXUUAwQJM8mfY5bW
▪▪▪▪▪ Done
```
2. the format of the airdrop_list.json file has not been added to `sugar aidrop -h`. Was this done on purpose?

After those two minor things I changed I would recommend to merge this PR!

**Comment by ohaddahan** [comment]
> 

Thank you, missed those https://github.com/metaplex-foundation/sugar/pull/375/commits/424fb7ec3a8521d9e9abc8ea2400a935de9fc6b1 .

**Review by samuelvanderwaal — APPROVED** [review]

**Pull request merged**